cli/src/main/java/rules/CheckTagsAreDefinedRule.java

```diff
@@ -0,0 +1,23 @@
+package rules;
+
+import org.eclipse.microprofile.openapi.models.OpenAPI;
+import org.openapitools.openapistylevalidator.ErrorAggregator;
+import org.openapitools.openapistylevalidator.custom.OpenAPIRule;
+import org.openapitools.openapistylevalidator.styleerror.StyleError;
+
+/**
+ * This rule will check whether tags are defined in the rule section
+ */
+public class CheckTagsAreDefinedRule extends OpenAPIRule {
```

Comment:
I think that if we go that route, we might need to convert ALL the validations / rules in the current code to this model. That way, we could stay backward compatible (for a while at least) while using this new "framework".

I also think that "custom rules" should be marked as so. They would be the "official" rules and the "custom". Maybe have a prefix "Custom" in the name of the class? What do you think?

Reply:
Yes.. we should move all the existing rules in the design and make it backward compatible. We can implement rules for common scenarios(i.e the current validation supported by the library) into the new framework and provide library for them to write their own rules by extending  `OpenAPIRule` abstract class.  If you look at the custom generator approach user will provide their library for their own generator in the class path and openapi-geneator is smart enough find the users custom generator from classpath. I want to do the same approach here as well.

Reply:
That's a very good idea but I think it could be very interesting to provide both.
A "library" of reusable rules that is built in, including the existing one + other that people request often.
But also, like you say, the possibility to dynamically load one in the classpath for extension.